File: src/Reference.VB/DatabaseService.vb
Imports Microsoft.EntityFrameworkCore
' ReSharper disable UnusedMember.Global

''' <summary>
''' Database service
''' </summary>
Public Class DatabaseService
    Implements IPersistentData

    ''' <inheritdoc />
    Public Async Function EnsureMigrationAsync() As Task Implements IPersistentData.EnsureMigrationAsync
        Using context = New GeoContext()
            Await context.Database.MigrateAsync()
        End Using
    End Function

    ''' <inheritdoc />
    Public Async Function SaveDataAsync(geoPoint As GeoPoint, item As OpenStreetMapItem) As Task(Of Boolean) Implements IPersistentData.SaveDataAsync
        Using context = New GeoContext()
            Dim entity = Await context.GeoData.FindAsync(geoPoint.Uuid)
            If entity IsNot Nothing Then
                Return False
            End If

            entity = New GeoEntity With {
                .Uuid = geoPoint.Uuid,
                .Latitude = geoPoint.Latitude,
                .Longitude = geoPoint.Longitude,
                .PlaceId = item.PlaceId,
                .DisplayName = item.DisplayName,
                .City = item.Address?.City,
                .Country = item.Address?.Country,
                .CountryCode = item.Address?.CountryCode,
                .County = item.Address?.County,
                .State = item.Address?.State
                }

            Await context.GeoData.AddAsync(entity)
            Return (Await context.SaveChangesAsync()) > 0
        End Using
    End Function

    ''' <inheritdoc />
    Public Async Function GetAllAsync() As Task(Of List(Of (GeoPoint As GeoPoint, StreetMap As OpenStreetMapItem))) Implements IPersistentData.GetAllAsync
        Dim lstResponse = New List(Of (GeoPoint As GeoPoint, StreetMap As OpenStreetMapItem))
        Using context = New GeoContext()
            Dim data = Await context.GeoData.ToArrayAsync()
            For Each item As GeoEntity In data
                Dim geoPoint = New GeoPoint() With {
                        .Uuid = item.Uuid,
                        .Latitude = item.Latitude,
                        .Longitude = item.Longitude
                        }
                Dim streetMap = New OpenStreetMapItem() With {
                        .PlaceId = item.PlaceId,
                        .DisplayName = item.DisplayName,
                        .Address = New OpenStreetMapAddress() With {
                            .City = item.City,
                            .Country = item.Country,
                            .CountryCode = item.CountryCode,
                            .County = item.County,
                            .State = item.State
                        }
                }
                lstResponse.Add((geoPoint, streetMap))
            Next
        End Using
        Return lstResponse
    End Function

End Class

File: src/Reference.CSharp/DatabaseService.cs
using Microsoft.EntityFrameworkCore;
using System.Collections.Generic;
using System.Threading.Tasks;
// ReSharper disable UnusedMember.Global

namespace Reference.CSharp
{
    /// <summary>
    /// Database service
    /// </summary>
    public class DatabaseService : IPersistentData
    {
        /// <inheritdoc />
        public async Task EnsureMigrationAsync()
        {
            using (var context = new GeoContext())
                await context.Database.MigrateAsync();
        }

        /// <inheritdoc />
        public async Task<bool> SaveDataAsync(GeoPoint geoPoint, OpenStreetMapItem item)
        {
            using (var context = new GeoContext())
            {
                var entity = await context.GeoData.FindAsync(geoPoint.Uuid);
                if (entity != null)
                    return false;

                entity = new GeoEntity
                {
                    Uuid = geoPoint.Uuid,
                    Latitude = geoPoint.Latitude,
                    Longitude = geoPoint.Longitude,
                    PlaceId = item.PlaceId,
                    DisplayName = item.DisplayName,
                    City = item.Address?.City,
                    Country = item.Address?.Country,
                    CountryCode = item.Address?.CountryCode,
                    County = item.Address?.County,
                    State = item.Address?.State
                };

                await context.GeoData.AddAsync(entity);
                return await context.SaveChangesAsync() > 0;
            }
        }

        /// <inheritdoc />
        public async Task<List<(GeoPoint GeoPoint, OpenStreetMapItem StreetMap)>> GetAllAsync()
        {
            var lstResponse = new List<(GeoPoint GeoPoint, OpenStreetMapItem StreetMap)>();
            using (var context = new GeoContext())
            {
                var data = await context.GeoData.ToArrayAsync();
                foreach (var item in data)
                {
                    lstResponse.Add((
                        new GeoPoint
                        {
                            Uuid = item.Uuid,
                            Latitude = item.Latitude,
                            Longitude = item.Longitude
                        }, 
                        new OpenStreetMapItem
                        {
                            PlaceId = item.PlaceId,
                            DisplayName = item.DisplayName,
                            Address = new OpenStreetMapAddress
                            {
                                City = item.City,
                                Country = item.Country,
                                CountryCode = item.CountryCode,
                                County = item.County,
                                State = item.State
                            }
                        }));
                }
            }
            return lstResponse;
        }
    }
}